netty/src/main/java/io/grpc/netty/Utils.java

```diff
@@ -168,26 +169,33 @@ private static Http2Headers convertMetadata(Metadata headers) {
   }
 
   public static Status statusFromThrowable(Throwable t) {
+    // First, inspect the causal chain explicitly looking for status exceptions.
     Status s = Status.fromThrowable(t);
     if (s.getCode() != Status.Code.UNKNOWN) {
+      // It was already a status exception, just use it.
       return s;
     }
-    // TODO(ejona): reenable once startup races are resolved; ClosedChannelException is being seen
-    // still. Some tests are asserting UNAVAILABLE and were "working" previously but are now
-    // detecting that our behavior is flaky. See #1330
-    if (false && t instanceof ClosedChannelException) {
-      // ClosedChannelException is used any time the Netty channel is closed. Proper error
-      // processing requires remembering the error that occurred before this one and using it
-      // instead.
-      return Status.UNKNOWN.withCause(t);
-    }
-    if (t instanceof IOException) {
-      return Status.UNAVAILABLE.withCause(t);
-    }
-    if (t instanceof Http2Exception) {
-      return Status.INTERNAL.withCause(t);
+
+    // Assume UNKNOWN unless we find a better status.
+    String msg = t.getMessage();
+    Status status = Status.UNKNOWN.withCause(t).withDescription(msg);
+
+    // Iterate over the causal chain for the error and find the best status.
+    for (Throwable cause : Throwables.getCausalChain(t)) {
+      if (cause instanceof IOException) {
+        // Map any IOExceptions to UNAVAILABLE to allow retry.
+        msg = msg != null ? msg : cause.getMessage();
+        return Status.UNAVAILABLE.withCause(t).withDescription(msg);
+      }
+
+      if (cause instanceof Http2Exception) {
+        // Use INTERNAL in case we can't find anything better.
+        msg = msg != null ? msg : cause.getMessage();
+        status = Status.INTERNAL.withCause(t).withDescription(msg);
```

Comment:
The rebuilding of the status each time is part of why I was uncomfortable with it. I'm still uncomfortable with it, but not as uncomfortable with it now that I see it uses `t` instead of `cause` as the Status's cause.

P.S. Uncheck "Show notes" at the top right of this file's diff (next to view, across from the file name).